src/middlewares/Cache.ts

```diff
@@ -0,0 +1,149 @@
+import {objectToQueryString} from "../internal/formatting";
+import {FetchHeaders, FetchRequest, FetchResponse} from "./Fetch";
+
+import {IMiddleware} from "../Stack";
+
+// Represents nature of cache entry that is persisted into the configured engine.
+export interface ICacheEntry {
+    value: string;      // text representation of response payload
+    type: string;       // the response type.
+    url: string;        // the url that responded original cache response
+    status: number;     // http response status
+    statusText: string; // http response status text. eg: OK
+    timestamp: number;  // create timestamp in milliseconds: +Date.now()
+    headers: {[index: string]: any}; // response headers
+}
+
+// Represents cache options for each cache buckets added to the cache middleware
+export interface IFetchRequestCacheOptions {
+    enable: boolean;
+    maxAge?: number; // cache ttl in seconds, defaults to 0 = forever
+    key?: string; // optionally a specific key to be used in place of the generated hash
+}
+
+// Represents the storage engine functionality
+export interface ICacheMiddlewareStore {
+    setItem(key: string, value: string): void;
+    getItem(key: string): string|undefined;
+    removeItem(key: string): void;
+}
+
+// noinspection TsLint
+declare const Response: FetchResponse<string>;
+
+export class Cache implements IMiddleware<FetchRequest, Promise<FetchResponse<any>>> {
+    protected fallbackStorage: {[index: string]: any} = {};
+
+    /**
+     * Constructor.
+     *
+     * Accepts storage instance.
+     *
+     * @param {ICacheMiddlewareStore|undefined} config
+     */
+    public constructor(protected storage: ICacheMiddlewareStore|undefined = undefined) {
+    }
+
+    /**
+     * Process the request. First try if we have cache and serve right away,
+     * else let the next middleware in pileline be invoked and cache it.
+     *
+     * @param  {FetchRequest} options
+     * @param  {callback}     next
+     * @return {any}
+     */
+    public process(options: FetchRequest, next: (nextOptions: FetchRequest) => Promise<FetchResponse<any>>): any {
+        // Cache not configured/enabled
+        if (!options.cache || !options.cache.enable) { return next(options); }
+
+        const key = this.getCacheKey(options);
+        let entry = this.storage && this.storage.getItem(key) || this.fallbackStorage[key];
+
+        /* istanbul ignore else */
```

Comment:
was actually needed for branching coverage as we dont have `else`, looks like no need now. strange!